neutrons/SNAPRed
# Ewm3143 apply diffcal during norm workflow

## Description of work

This supplies the features requested in ewm3143:
- The application of a diffraction calibration to the normaliztion data in the Normalization Workflow for user assessment
- The logging of metadata on the workspace if it uses the default calibration in the Normalization Workflow

This also supplies features in the neighborhood of what it asked for: 
- The logging of metadata on the workspaces if diffraction calibration/normalization is missing in the Reduction Workflow

I have also refactored:
- `StrEnum` to live in a seperate `Enum` module.  
- `WorkspaceMetadata` now uses `StrEnum` instead of literals.
- Recipes now can supply an init method for their given Ingredients.

## To test

NOTE:  Use the View Sample Logs option when right clicking on a workspace to confirm it has the snapred metadata.

```
For reduction
Runnumber: 59039 

for calib:
sample: LA11b6
Grouping: Column
Peak Intensity Thresh: 0.01
Chi: 200
dmin: 1

for norm:
rnunumber: 58810
bgrunnumber: 58813
```
Attempt to run either [Normalization Worflow, Reduction Workflow] -> [without a state, with a default state, with a valid state]
### Normalization
#### without a state

1. should return a runtime error if you lack permission to init state
2. should ask you to init state if you have write permission

#### with a default state
1. Should prompt with a continue anyway informing the user that the output will be marked as potentially inaccurate by means of it using a default state.

#### with a valid state
1. Continue as normal

### Reduction

As it is in `next` except:
- it actually loads the diffcal mask
- the outputs should have appropriately marked metadata

## Link to EWM item


[EWM#3143](https://ornlrse.clm.ibmcloud.com/ccm/web/projects/Neutron%20Data%20Project%20(Change%20Management)#action=com.ibm.team.workitem.viewWorkItem&id=3143)
[EWM#7758](https://ornlrse.clm.ibmcloud.com/ccm/web/projects/Neutron%20Data%20Project%20(Change%20Management)#action=com.ibm.team.workitem.viewWorkItem&id=7758)

### Verification

- [x] the author has read the EWM story and acceptance critera (Malcolm has additionally review and asked me to add AC)


### Acceptance Criteria

This list is for ease of reference, and does not replace reading the EWM story as part of the review.  Verify this list matches the EWM story before reviewing.

~~Unfortunately there is not an explicity Acceptance Criteria outlined in the story, what I can infer from it are as follows~~
Malcolm has since reviewed the AC and made clarifications.

- [x] in the normalisation workflow, the raw vanadium data shall be treated *exactly the same* as the sample data being Reduced for examination.
- [x] in the normalisation workflow, If there is only a default diffcal (This workflow requires at least an initialized state), there should be a warning that the output will be labeled as using such default diffcal and proceed anyway prompt.
- [x] The outputs should have a log on them denoting the existence/status of diffcal/normalization.
- [ ] in the normalisation workflow, the raw vanadium data shall be treated exactly the same as the sample data being Reduced for examination.
- [x] the difcal index is queried to find available difcal for the normalization run 
- [x] if found the diffcal is applied to the raw 
- [x] Any corresponding diffcal mask is also applied to the raw vanadium

7758
- [x]  The correct behaviour would be to warn the user that the state does not exist, prompt user to create the state (probably with a notice that only an instrument scientist can actually do this)
- [x] if state is initialized as above and the user elects to proceed with default calibration, it should not fail trying to load such calibration
